seq_tools.py

```diff
@@ -1,102 +1,333 @@
+
+    def validate_protein(self) -> bool:
+        """
+        Raises:
+            ValueError: in case sequence is empty 
+            ValueError: in case characters in sequence are not allowed
+            ValueError: in case sequence does not start from methionine 
+            ValueError: in case sequence us too short to form a protein
+
+        Returns:
+            bool: True if the sequence is a valid protein, otherwise raises an exception
+        """
+         
+        if not self.sequence:
+            raise ValueError("Sequence is empty")
+        if not self._check_alphabet():
+            raise ValueError("")
+        if self.sequence[0] != self._start_codon:
+            raise ValueError("Sequence does not start with methionine (M)")
+        if len(self.sequence) < 20:
+            raise ValueError("Sequence is too short to be a valid protein")
+        return True
+
+
+def filter_fastq(
+    input_file: str,
+    output_file: str = "output_filter.fastq",
+    gc_bounds: tuple[float, float] = (0, 100),
+    length_bounds: tuple[int, int] = (0, 2**32),
+    quality_threshold: float = 0,
+) -> None:
+    """Filter FASTQ sequences according to length, GC content, and quality thresholds.
+    The result is saved in an output file in a directory named 'filtered'.
+
+    Args:
+        input_file (str): Path to the input FASTQ file.
+        output_file (str, optional): Output file name with extension.
+            Defaults to 'output_filter.fastq'.
+        gc_bounds (tuple[float, float], optional): GC content bounds (min, max).
+            Defaults to (0, 100).
+        length_bounds (tuple[int, int], optional): Length bounds (min, max).
+            Defaults to (0, 2**32).
+        quality_threshold (float, optional): Minimum average quality score.
+            Defaults to 0.
+    """
+    output_dir = "filtered"
+    os.makedirs(output_dir, exist_ok=True)
+    output_path = os.path.join(output_dir, output_file)
+
+    filtered_records = []
+    for record in SeqIO.parse(input_file, "fastq"):
+        if not record.seq:
             continue
-        if not na_type(seq):
-            ans.append(None)
-            stderr[seq] = (na_type(seq), "RNA and DNA mix")
-        else:
-            if task != "transcribe":
-                ans.append(operations.get(task)(seq))
-            elif task == "transcribe" and na_type(seq) == "DNA":
-                ans.append(operations.get(task)(seq))
-            else:
-                ans.append(None)
-                stderr[seq] = (
-                    0,
-                    "operation type is not compatible with NA type",
-                )
-    if stderr:
-        for seq, (result, message) in stderr.items():
-            print(f" {seq} : {message}")
-    if len(ans) == 1:
-        return ans[0]
-    return ans
+
+        gc_content = gc_fraction(record.seq) * 100
+        avg_quality = sum(record.letter_annotations["phred_quality"]) / len(
+            record
+        )
```

Comment:
Я бы тогда так сделал:

```suggestion
        qualities = record.letter_annotations["phred_quality"]
        avg_quality = sum(qualities ) / len(record)
```
Лучше чем разбивать расчет длины на 3 строки